website/content/ChapterFour/0001~0099/0015.3Sum.md

```diff
@@ -44,35 +44,24 @@ import (
 
 // 解法一 最优解，双指针 + 排序
 func threeSum(nums []int) [][]int {
-	sort.Ints(nums)
-	result, start, end, index, addNum, length := make([][]int, 0), 0, 0, 0, 0, len(nums)
-	for index = 1; index < length-1; index++ {
-		start, end = 0, length-1
-		if index > 1 && nums[index] == nums[index-1] {
-			start = index - 1
-		}
-		for start < index && end > index {
-			if start > 0 && nums[start] == nums[start-1] {
-				start++
-				continue
-			}
-			if end < length-1 && nums[end] == nums[end+1] {
-				end--
-				continue
-			}
-			addNum = nums[start] + nums[end] + nums[index]
-			if addNum == 0 {
-				result = append(result, []int{nums[start], nums[index], nums[end]})
-				start++
-				end--
-			} else if addNum > 0 {
-				end--
-			} else {
-				start++
-			}
-		}
-	}
-	return result
+    answer := [][]int{}
+    numsIdxMap := map[int]int{}
+    l := len(nums)
+    sort.Ints(nums)
+    for i, v := range nums {
+        numsMap[v] = i
+    }
+    for i := 0; i < l ; i ++ {
+        for j := i + 1 ; j < l && nums[j] <= -nums[i]; j++ {
```

Comment:
There are two-layer loop, so this solution time complexity is O(n^2). My solution time complexity is O(n log n). From a runtime perspective, this solution runtime is [Runtime: 52 ms](https://leetcode.com/submissions/detail/490331402/), my solution runtime is [Runtime: 24 ms](https://leetcode.com/submissions/detail/449571779/).